### PR TITLE
Charge and transfer by date

### PR DESCRIPTION
**Pull request by VovaViter**

### CODE REVIEW

File: src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java [hunk 1]
```diff
@@ -281,17 +282,18 @@ public TransferDto transfer(TransferBalanceRequest transferRequest) {
         Balance sourceBalance = getBalanceForUpdate(transferRequest.getSourceBalanceId());
         BigDecimal amount = transferRequest.getAmount();
         Metadata metadata = Metadata.of(transferRequest.getMetadata());
-        assertBalanceAmount(sourceBalance, amount);
 
-        Instant now = now(clock);
+        Instant now = transferRequest.getApplyDate() == null ?  now(clock) : transferRequest.getApplyDate();
+
+        assertBalanceAmount(sourceBalance, amount);
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java]: pls apply date logic to BalanceRepository.findBalanceAmount too

File: src/main/java/com/icthh/xm/ms/balance/web/rest/requests/ChargingBalanceRequest.java [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package com.icthh.xm.ms.balance.web.rest.requests;
 
 import java.math.BigDecimal;
+import java.time.Instant;
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/balance/web/rest/requests/ChargingBalanceRequest.java:4]: unused import

File: src/main/java/com/icthh/xm/ms/balance/web/rest/BalanceResource.java [hunk 3]
```diff
@@ -30,14 +31,7 @@
...
 import org.springframework.security.access.prepost.PreAuthorize;
-import org.springframework.web.bind.annotation.DeleteMapping;
-import org.springframework.web.bind.annotation.GetMapping;
-import org.springframework.web.bind.annotation.PathVariable;
-import org.springframework.web.bind.annotation.PostMapping;
-import org.springframework.web.bind.annotation.PutMapping;
-import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.*;
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/balance/web/rest/BalanceResource.java]: avoid to user *, pls revert this change

File: src/main/java/com/icthh/xm/ms/balance/web/rest/BalanceResource.java [hunk 4]
```diff
@@ -127,8 +121,8 @@ public BalanceResource(BalanceService balanceService, BalanceQueryService balanc
     @GetMapping("/balances/{id}")
     @Timed
     @PrivilegeDescription("Privilege to get the balance by id")
-    public ResponseEntity<BalanceDTO> getBalance(@PathVariable Long id) {
-        BalanceDTO balanceDTO = balanceService.findOne(id);
+    public ResponseEntity<BalanceDTO> getBalance(@PathVariable Long id, @RequestParam(value = "applyDate", required = false) Instant applyDate) {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/balance/web/rest/BalanceResource.java:131]: pls check how it work when this parameter not passed

File: src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java [hunk 5]
```diff
@@ -122,11 +122,11 @@ public BalanceDTO save(BalanceDTO balanceDTO) {
      * @return the entity
      */
     @Transactional(readOnly = true)
-    public BalanceDTO findOne(Long id) {
+    public BalanceDTO findOne(Long id, Instant applyDate) {
         Balance balance = balanceRepository.findById(id).orElse(null);
         BalanceDTO balanceDTO = balanceMapper.toDto(balance);
         if (balanceDTO != null) {
-            balanceDTO.setAmount(balanceRepository.findBalanceAmount(balance).orElse(ZERO));
+            balanceDTO.setAmount(balanceRepository.findBalanceAmount(balance, applyDate == null ? now(clock) : applyDate).orElse(ZERO));
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java:129]: так а если applyDate не передали?

Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/balance/service/BalanceService.java]: select variable

File: src/test/java/com/icthh/xm/ms/balance/repository/BalanceRepositoryIntTest.java [hunk 6]
```diff
@@ -27,7 +28,7 @@
     @DataSet(value = "mockBalances-init.xml", disableConstraints = true)
     public void amountCalculatedFromPockedWithFilterByDate() {
```
Comment (sergeysenja1992) [inline src/test/java/com/icthh/xm/ms/balance/repository/BalanceRepositoryIntTest.java:30]: pls add test that check date change request result

File: src/test/java/com/icthh/xm/ms/balance/repository/PocketRepositoryIntTest.java [hunk 7]
```diff
@@ -53,17 +55,17 @@ public void amountCalculatedFromPockedWithFilterByDateLimit() {
         List<Long> expectedOrder = asList(11L, 18L, 24L, 13L, 22L, 25L, 23L, 14L, 15L, 16L, 19L, 17L, 20L, 21L);
 
         Page<Pocket> pockets = pocketRepository.findPocketForChargingOrderByDates(balanceRepository.findById(4L).get(),
```
Comment (sergeysenja1992) [inline src/test/java/com/icthh/xm/ms/balance/repository/PocketRepositoryIntTest.java:57]: add test that check is date affect result fo request

File: src/test/java/com/icthh/xm/ms/balance/service/BalanceServiceUnitTest.java [hunk 8]
```diff
@@ -72,6 +71,7 @@
 public class BalanceServiceUnitTest {
 
     private static final String EMPTY_METADATA_VALUE = null;
+    private static final Instant APPLY_DATE_VALUE = Instant.ofEpochMilli(1525428386000L);
```
Comment (sergeysenja1992) [inline src/test/java/com/icthh/xm/ms/balance/service/BalanceServiceUnitTest.java]: rename to MOCK_CURRENT_DATE

**Review by sergeysenja1992 — APPROVED** [review]

**Review by amedvedchuk — APPROVED** [review]
